File: Leetcode/June Challenge/day 5/Random Pick with Weight.cpp
class Solution {
public:
    vector<int>a;
    int n;
    Solution(vector<int>& w) {
        n=w[0];
        for(int i=1;i<w.size();i++)
        {
            w[i]+=w[i-1];
            n=w[i];
        }
        a=w;
        
    }
    
    int pickIndex() {
        int r =rand();
        r=r%n;
        for(int i=0;i<a.size();i++)
        {
            if(r<a[i])
                return i;
        }
        return NULL;
        
    }
};

/**
 * Your Solution object will be instantiated and called as such:
 * Solution* obj = new Solution(w);
 * int param_1 = obj->pickIndex();
 */

File: Leetcode/June Challenge/day 5/Random Pick with Weight.java
class Solution {
    int []arr;
    int n;
    public Solution(int[] w) {
        arr=new int [w.length];
        arr[0]=w[0];
        for(int i=1;i<w.length;i++)
        {
            w[i]=w[i-1]+w[i];
            arr[i]=w[i];
        }
        int index=w.length-1;
        n=arr[index];
    }
    
    public int pickIndex() {
        double rand=Math.random()*n;
        for(int i=0;i<arr.length;i++)
        {
            if(rand<arr[i])
            {
                return i;
            }
        }
        return arr.length-1;
    }
}

/**
 * Your Solution object will be instantiated and called as such:
 * Solution obj = new Solution(w);
 * int param_1 = obj.pickIndex();
 */